File: common/python/argo/template/v1/deployment.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2015-2017 Applatix, Inc. All rights reserved.
#

import re

from future.utils import iteritems

from .base import BaseTemplate
from .common import Inputs, TerminationPolicy
from .container import ContainerTemplate
from argo.parser import ObjectParser
from argo.parser import ObjectDefinition as OD
from argo.template.v1.common import VolumeDetails


class DeploymentTemplate(BaseTemplate):

    """
    :type inputs: Inputs
    :type scale: Scale
    """

    def __init__(self):
        super(DeploymentTemplate, self).__init__()
        self.inputs = Inputs()
        self.application_name = None
        self.deployment_name = None
        self.scale = Scale()
        self.external_routes = []
        self.internal_routes = []
        self.containers = {}
        self.fixtures = []
        self.volumes = {}
        self.termination_policy = None
        self.min_ready_seconds = None
        self.strategy = Strategy()

        fields = {
            "inputs": Inputs.fqcn(),
            "application_name": None,
            "deployment_name": None,
            "scale": Scale.fqcn(),
            "external_routes": ExternalRoute.fqcn(),
            "internal_routes": InternalRoute.fqcn(),
            "containers": None, # needs special handling later
            "fixtures": None,
            "volumes": None, # needs special handling later
            "termination_policy": TerminationPolicy.fqcn(),
            "min_ready_seconds": None,
            "strategy": Strategy.fqcn()
        }

        self._object_parser_fields.update(fields)

    def parse(self, data, error_on_not_found=False):
        super(DeploymentTemplate, self).parse(data, error_on_not_found=error_on_not_found)

        # now handle containers
        for k,v in iteritems(self.containers):
            new_container = ContainerTemplate()
            if "template" in v:
                # found the service object not template
                new_container.parse(v["template"], error_on_not_found=error_on_not_found)
            else:
                new_container.parse(v, error_on_not_found=error_on_not_found)
            self.containers[k] = new_container

        # now handle volumes
        for k,v in iteritems(self.volumes):
            new_vol = VolumeDetails()
            new_vol.parse(v, error_on_not_found=error_on_not_found)
            self.volumes[k] = new_vol

    def to_dict(self):
        ret = super(DeploymentTemplate, self).to_dict()

        # handle special cases now
        ret["containers"] = {}
        for k,v in iteritems(self.containers):
            ret["containers"][k] = v.to_dict()

        ret["volumes"] = {}
        for k,v in iteritems(self.volumes):
            ret["volumes"][k] = v.to_dict()

        return ret

    def get_main_container(self):
        for name, template in iteritems(self.containers):
            if template.name is None:
                template.name = name
            return template
        raise ValueError("There needs to be at least one container defined in the specification")


class Scale(ObjectParser):

    def __init__(self):
        self.min = 1
        self.max = None

        self._object_parser_fields = {
            "min": None,
            "max": None
        }
        super(Scale, self).__init__()


class ExternalRoute(ObjectParser):

    def __init__(self):
        self.name = None
        self.dns_name = None
        self.dns_prefix = None
        self.dns_domain = None
        self.target_port = None
        self.ip_white_list = []
        self.visibility = None

        self._object_parser_fields = {
            "name": None,
            "dns_name": DnsDomain.fqcn(),
            "dns_prefix": DnsLabel.fqcn(),
            "dns_domain": DnsDomain.fqcn(),
            "target_port": None,
            "ip_white_list": None,
            "visibility": None
        }
        super(ExternalRoute, self).__init__()


class DnsLabel(ObjectParser):
    """
    This Subclass of ObjectParser parses dns label and matches them with
    regex to ensure that it is a valid dns label
    """

    def __init__(self):
        self.dns_label = None
        # don't need _object_parser_fields as it has no keys that it needs to parse
        super(DnsLabel, self).__init__()

    def parse(self, data, error_on_not_found=False):
        """
        regex is [a-z0-9]([-a-z0-9]{0,61}[a-z0-9])?
        """
        matched = re.match(r"^[a-z0-9]([-a-z0-9]{0,61}[a-z0-9])?$", data.lower())
        if not matched:
            raise ValueError("{} is not a valid Dns Label".format(data))
        self.dns_label = data.lower()

    def __call__(self, *args, **kwargs):
        return self.dns_label

    def to_dict(self):
        return self.dns_label


class DnsDomain(ObjectParser):

    def __init__(self):
        self.dns_domain = None
        super(DnsDomain, self).__init__()
    
    def parse(self, data, error_on_not_found=False):
        labels = data.lower().split(".")
        for label in labels:
            if len(label) == 0:
                continue
            matched = re.match(r"^[a-z0-9]([-a-z0-9]{0,61}[a-z0-9])?$", label)
            if not matched:
                raise ValueError("Dns Domain {} is not valid as the label {} did not match requirements".format(label, data.lower()))
        
        self.dns_domain = data.lower()
    
    def __call__(self, *args, **kwargs):
        return self.dns_domain

    def to_dict(self):
        return self.dns_domain


class InternalRoute(ObjectParser):

    def __init__(self):
        self.name = None
        self.ports = []

        self._object_parser_fields = {
            "name": None,
            "ports": Port.fqcn()
        }
        super(InternalRoute, self).__init__()


class Port(ObjectParser):

    def __init__(self):
        self.port = None
        self.target_port = None

        self._object_parser_fields = {
            "port": None,
            "target_port": None
        }
        super(Port, self).__init__()


class Strategy(ObjectParser):

    def __init__(self):
        self.type = "recreate"
        self.rolling_update = RollingUpdateStrategy()

        self._object_parser_fields = {
            "type": None,
            "rolling_update": RollingUpdateStrategy.fqcn()
        }
        super(Strategy, self).__init__()


class RollingUpdateStrategy(ObjectParser):

    def __init__(self):
        self.max_surge = 1
        self.max_unavailable = 1

        self._object_parser_fields = {
            "max_surge": OD(converter_tuple=(int, str)),
            "max_unavailable": OD(converter_tuple=(int, str))
        }
        super(RollingUpdateStrategy, self).__init__()

File: saas/common/src/applatix.io/template/deployment.go
package template

import (
	"strconv"
	"strings"

	"applatix.io/axerror"
	"applatix.io/common"
)

const (
	// External route options
	ExternalRouteVisibilityWorld = "world"
	ExternalRouteVisibilityOrg   = "organization"
	// Upgrade Strategies
	StrategyRollingUpdate = "rolling_update"
	StrategyRecreate      = "recreate"
)

// NOTE: squash is a mapstructure struct tag but we teach mapstructure to parse the json tags
type DeploymentTemplate struct {
	BaseTemplate      `json:",squash"`
	Inputs            *Inputs                               `json:"inputs,omitempty"`
	ApplicationName   string                                `json:"application_name,omitempty"`
	DeploymentName    string                                `json:"deployment_name,omitempty"`
	Scale             *Scale                                `json:"scale,omitempty"`
	ExternalRoutes    ExternalRoutes                        `json:"external_routes,omitempty"`
	InternalRoutes    InternalRoutes                        `json:"internal_routes,omitempty"`
	Containers        map[string]InlineContainerTemplateRef `json:"containers,omitempty"`
	Fixtures          FixtureRequirements                   `json:"fixtures,omitempty"`
	Volumes           VolumeRequirements                    `json:"volumes,omitempty"`
	TerminationPolicy *TerminationPolicy                    `json:"termination_policy,omitempty"`
	MinReadySeconds   int                                   `json:"min_ready_seconds,omitempty"`
	Strategy          *Strategy                             `json:"strategy,omitempty"`
}

type Scale struct {
	Min int `json:"min,omitempty"`
	Max int `json:"max,omitempty"`
}

type ExternalRoutes []*ExternalRoute

type ExternalRoute struct {
	DNSPrefix   string   `json:"dns_prefix,omitempty"`
	DNSDomain   string   `json:"dns_domain,omitempty"`
	DNSName     string   `json:"dns_name,omitempty"`
	TargetPort  string   `json:"target_port,omitempty"`
	IPWhiteList []string `json:"ip_white_list,omitempty"`
	Visibility  string   `json:"visibility,omitempty"`
}

type InternalRoutes []*InternalRoute

type InternalRoute struct {
	Name  string  `json:"name,omitempty"`
	Ports []*Port `json:"ports,omitempty"`
}

type Port struct {
	Port       string `json:"port,omitempty"`
	TargetPort string `json:"target_port,omitempty"`
}

type Strategy struct {
	Type          string                 `json:"type,omitempty"`
	RollingUpdate *RollingUpdateStrategy `json:"rolling_update,omitempty"`
}

type RollingUpdateStrategy struct {
	MaxSurge       string `json:"max_surge"`
	MaxUnavailable string `json:"max_unavailable"`
}

func (tmpl *DeploymentTemplate) GetInputs() *Inputs {
	return tmpl.Inputs
}

func (tmpl *DeploymentTemplate) GetOutputs() *Outputs {
	return nil
}

func (tmpl *DeploymentTemplate) Validate(preproc ...bool) *axerror.AXError {
	preprocessing := len(preproc) > 0 && preproc[0]
	if tmpl.Name == "" {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessage("'name' field missing or empty")
	}
	// TODO: validate kubernetes name
	tmpl.ApplicationName = strings.TrimSpace(tmpl.ApplicationName)
	if tmpl.ApplicationName == "" {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessage("'application' field missing or empty")
	}
	tmpl.DeploymentName = strings.TrimSpace(tmpl.DeploymentName)
	if tmpl.DeploymentName == "" {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessage("'deployment' field missing or empty")
	}
	axErr := tmpl.Inputs.Validate(false)
	if axErr != nil {
		return axErr
	}
	axErr = tmpl.Fixtures.Validate()
	if axErr != nil {
		return axErr
	}
	if tmpl.Scale == nil {
		tmpl.Scale = &Scale{
			Min: 1,
		}
	}
	if tmpl.Volumes != nil {
		if len(tmpl.Volumes) > 0 && tmpl.Scale.Min > 1 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("%s.scale: deployment requiring volumes cannot scale past 1 instance", tmpl.Name)
		}
		axErr := tmpl.Volumes.Validate()
		if axErr != nil {
			return axErr
		}
	}
	if tmpl.InternalRoutes != nil {
		axErr := tmpl.InternalRoutes.Validate()
		if axErr != nil {
			return axErr
		}
	}
	if tmpl.ExternalRoutes != nil {
		axErr := tmpl.ExternalRoutes.Validate()
		if axErr != nil {
			return axErr
		}
	}
	if tmpl.Fixtures != nil {
		for i, parallelFixtures := range tmpl.Fixtures {
			for fixRefName, ftr := range parallelFixtures {
				if ftr.IsDynamicFixture() {
					return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("fixtures[%d].%s: deployments cannot use dynamic fixtures", i, fixRefName)
				}
			}
		}
	}
	if tmpl.Strategy != nil && tmpl.Strategy.Type == StrategyRollingUpdate {
		if len(tmpl.Volumes) > 0 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy type \"rolling_update\" is not supported for deployments with volume")
		}
	}
	// run Validate against any inlined containers
	for refName, ctRef := range tmpl.Containers {
		if ctRef.Inlined() {
			axErr := ctRef.Validate()
			if axErr != nil {
				return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("containers.%s: %v", refName, axErr)
			}
			_, _, axErr = ctRef.ReverseInline("temp")
			if axErr != nil {
				return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("containers.%s: %v", refName, axErr)
			}
		}
	}

	if !preprocessing {
		// Container checking is skipped if we are validating during preprocessing, because
		// in the context of a EmbeddedDeploymentTemplate, the containers field will be null.
		// This happens because containers field is "overwritten" by the child and becomes
		// dropped during marshalling/unmarshalling
		if len(tmpl.Containers) == 0 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessage("'containers' field missing or empty")
		}
		if len(tmpl.Containers) > 1 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessage("Only one container per deployment is currently supported")
		}

		axErr := tmpl.ValidateParameterScope()
		if axErr != nil {
			return axErr
		}
	}

	return nil
}

// ValidateParameterScope checks that all used parameters within the scope of the template, are declared and of the same type
func (tmpl *DeploymentTemplate) ValidateParameterScope() *axerror.AXError {
	declaredParams := tmpl.parametersInScope()
	usedParams, axErr := tmpl.usedParameters()
	if axErr != nil {
		return axErr
	}
	axErr = validateParams(declaredParams, usedParams)
	if axErr != nil {
		return axErr
	}
	return nil
}

func (tmpl *DeploymentTemplate) ValidateContext(context *TemplateBuildContext) *axerror.AXError {
	// This loop iterates the containers and checks if it is a template reference (as opposed to inlined container).
	// If it is, ensures we satisfy all parameters to the template, and ensure they are of the same type
	for refName, ctRef := range tmpl.Containers {
		if ctRef.Inlined() {
			continue
		}
		// Case where we are referencing another template
		st, ok := context.Templates[ctRef.Template]
		if !ok {
			return axerror.ERR_AXDB_INVALID_PARAM.NewWithMessagef("containers.%s.template: '%s' does not exist", refName, ctRef.Template)
		}
		if st.GetType() != TemplateTypeContainer {
			return axerror.ERR_AXDB_INVALID_PARAM.NewWithMessagef("containers.%s.template: '%s' must be of type container", refName, ctRef.Template)
		}

		ct := st.(*ContainerTemplate)
		scopedParams := tmpl.parametersInScope()
		axErr := validateReceiverParams(ct.Name, ct.Inputs, ctRef.Arguments, scopedParams)
		if axErr != nil {
			return axerror.ERR_AXDB_INVALID_PARAM.NewWithMessagef("containers.%s: %v", refName, axErr)
		}
	}
	return nil
}

func (tmpl *DeploymentTemplate) parametersInScope() paramMap {
	return getParameterDeclarations(tmpl.Inputs, tmpl.Volumes, tmpl.Fixtures)
}

// usedParameters detects all the parameters used in various parts of a deployment template and returns it in a paramMap.
// For deployments, check inline container's command, args, env, image
func (tmpl *DeploymentTemplate) usedParameters() (paramMap, *axerror.AXError) {
	pMap := make(paramMap)
	axErr := pMap.extractUsedParams(tmpl.ApplicationName, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.DeploymentName, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.ExternalRoutes, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.InternalRoutes, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.Fixtures, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.Volumes, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	axErr = pMap.extractUsedParams(tmpl.TerminationPolicy, paramTypeString)
	if axErr != nil {
		return nil, axErr
	}
	// merge any used parameters in inlined containers
	for refName, ct := range tmpl.Containers {
		ctParams, axErr := ct.usedParameters()
		if axErr != nil {
			return nil, axerror.ERR_API_INVALID_PARAM.NewWithMessagef("containers.%s: %v", refName, axErr)
		}
		axErr = pMap.merge(ctParams)
		if axErr != nil {
			return nil, axerror.ERR_API_INVALID_PARAM.NewWithMessagef("containers.%s: %v", refName, axErr)
		}
	}
	return pMap, nil
}

func (irs InternalRoutes) Validate() *axerror.AXError {
	set := map[string]bool{}
	for i, route := range irs {
		if axErr := route.Validate(); axErr != nil {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("internal_routes[%d]: %v", i, axErr)
		}
		if _, ok := set[route.Name]; ok {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("internal_routes[%d]: duplicated internal route name: %s", i, route.Name)
		}
		set[route.Name] = true
	}
	return nil
}

func (ir *InternalRoute) Validate() *axerror.AXError {
	if ir.Name == "" {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessage("'name' is required")
	}

	if !common.ValidateKubeObjName(ir.Name) {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'name' %v is not valid", ir.Name)
	}

	if len(ir.Ports) == 0 {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessage("no 'ports' provided")
	}

	for _, port := range ir.Ports {
		if _, err := strconv.ParseInt(port.Port, 0, 64); err != nil {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'port' %v is not valid: %v", port.Port, err.Error())
		}

		if _, err := strconv.ParseInt(port.TargetPort, 0, 64); err != nil {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'target_port' %v is not valid: %v", port.TargetPort, err.Error())
		}
	}

	return nil
}

func (ers ExternalRoutes) Validate() *axerror.AXError {
	dupPorts := make(map[string]bool)
	for i, route := range ers {
		if axErr := route.Validate(); axErr != nil {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("external_routes[%d]: %v", i, axErr)
		}
		if _, ok := dupPorts[route.TargetPort]; ok {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("external_routes[%d]: duplicated port: %s", i, route.TargetPort)
		}
		if len(route.DNSName) > 2000 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("external_routes[%d]: dns_name '%s' can not be longer than 2000 characters")
		}
		dupPorts[route.TargetPort] = true
	}
	return nil
}

func (er *ExternalRoute) Validate() *axerror.AXError {
	if _, err := strconv.ParseInt(er.TargetPort, 0, 64); err != nil {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'port' %v is not valid: %v", er.TargetPort, err.Error())
	}

	if er.IPWhiteList == nil || len(er.IPWhiteList) == 0 {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'ip_white_list' is required")
	}

	list := []string{}
	for _, ip := range er.IPWhiteList {
		list = append(list, strings.TrimSpace(ip))
	}

	er.IPWhiteList = list
	for _, ip := range er.IPWhiteList {
		if !common.ValidateCIDR(ip) {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("'ip_white_list' (%v) doesn't comply with IPv4 CIDR format", ip)
		}
	}

	if len(er.Visibility) == 0 {
		er.Visibility = ExternalRouteVisibilityWorld
	} else {
		if er.Visibility != ExternalRouteVisibilityWorld && er.Visibility != ExternalRouteVisibilityOrg {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("invalid external route visibility: '%s'. Valid values are: world, organization", er.Visibility)
		}
	}
	return nil
}

func (s *Strategy) Validate() *axerror.AXError {
	// validation for update strategy
	if s == nil {
		return nil
	}
	if s.Type != StrategyRollingUpdate && s.Type != StrategyRecreate {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy.type can only be one of [recreate, rolling_update]")
	}
	if s.Type == StrategyRecreate && s.RollingUpdate != nil {
		return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy type \"recreate\" cannot include configuration for rolling_update")
	}

	if s.RollingUpdate != nil {
		maxUnavailable := 1
		maxSurge := 1
		var err error
		if len(s.RollingUpdate.MaxUnavailable) > 0 {
			if maxUnavailable, err = strconv.Atoi(s.RollingUpdate.MaxUnavailable); err != nil {
				return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy.rolling_update.max_unavailable should be a number")
			}
		}
		if len(s.RollingUpdate.MaxSurge) > 0 {
			if maxSurge, err = strconv.Atoi(s.RollingUpdate.MaxSurge); err != nil {
				return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy.rolling_update.max_surge should be a number")
			}
		}
		if maxUnavailable == 0 && maxSurge == 0 {
			return axerror.ERR_API_INVALID_PARAM.NewWithMessagef("strategy.rolling_update.max_unavailable & strategy.rolling_update.max_surge cannot be 0 at the same time")
		}
	}

	return nil
}
